components/support/sql/src/open_database.rs

```diff
@@ -142,6 +195,72 @@ pub fn open_memory_database_with_flags<CI: ConnectionInitializer>(
     open_database_with_flags(":memory:", flags, conn_initializer)
 }
 
+// Handling a database failure after it is successfully opened means we should
+// probably try and drop the connection/transaction before attempting the
+// delete. This enum captures the various states we may be in. (You'd think it
+// could be simplified given Transaction impls Deref into a &Connection, but
+// markh can't work out how to leverage that)
+enum DroppableConnection<'a> {
+    None,                         // we failed attempting to get a connection.
+    Connection(Connection),       // we failed after opening but before our transaction.
+    Transaction(Transaction<'a>), // we failed after getting the transaction.
+}
+
+fn handle_db_failure<CI: ConnectionInitializer>(
+    open_info: OpenInfo<'_, CI>,
+    conn: DroppableConnection<'_>,
+    err: Error,
+) -> Result<Connection> {
+    if !open_info.handle_fatal_errors {
+        return Err(err);
+    }
+    log::warn!("{}: database operation failed: {}", CI::NAME, err);
+    if !open_info
+        .open_flags
+        .contains(OpenFlags::SQLITE_OPEN_READ_WRITE)
+    {
+        log::warn!(
+            "{}: not attempting recovery as this is a read-only connection request",
+            CI::NAME
+        );
+    }
+
+    let delete = match err {
+        Error::SqlError(RusqliteError::SqliteFailure(e, _)) => {
+            matches!(e.code, ErrorCode::DatabaseCorrupt | ErrorCode::NotADatabase)
+        }
+        _ => false,
+    };
+    if delete {
+        log::info!(
+            "{}: the database is fatally damaged; deleting and starting fresh",
+            CI::NAME
+        );
+        match conn {
+            DroppableConnection::None => {}
+            DroppableConnection::Connection(c) => drop(c),
+            DroppableConnection::Transaction(tx) => {
+                let _ = tx.rollback(); // ignore errors here.
+                                       // ideally we'd also call close the connection behind the transaction but
+                                       // that seems impossible - Defef into a `&Connection` fails here.
+            }
+        }
+        // XXX - should we move the path to, say ".corrupt"? Not clear there's
```

Comment:
yeah I can't think of a reason of keeping it unless we plan on maybe doing some diagnostics in a future time (then a log dump might also suffice)